pkg/envoy/secretsync.go

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package envoy
+
+import (
+	"context"
+	"fmt"
+
+	envoy_config_core_v3 "github.com/cilium/proxy/go/envoy/config/core/v3"
+	envoy_entensions_tls_v3 "github.com/cilium/proxy/go/envoy/extensions/transport_sockets/tls/v3"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	tlsCrtAttribute = "tls.crt"
+	tlsKeyAttribute = "tls.key"
+
+	// Key for CA certificate is fixed as 'ca.crt' even though is not as "standard"
+	// as 'tls.crt' and 'tls.key' are via k8s tls secret type.
+	caCrtAttribute = "ca.crt"
+)
+
+// secretSyncer is responsible to sync K8s TLS Secrets in pre-defined namespaces
+// via xDS SDS to Envoy.
+type secretSyncer struct {
+	logger         logrus.FieldLogger
+	envoyXdsServer XDSServer
+
+	mutex   lock.Mutex
+	secrets map[resource.Key]*slim_corev1.Secret
+}
+
+func newSecretSyncer(logger logrus.FieldLogger, envoyXdsServer XDSServer) *secretSyncer {
+	return &secretSyncer{
+		logger:         logger,
+		envoyXdsServer: envoyXdsServer,
+		secrets:        map[resource.Key]*slim_corev1.Secret{},
+	}
+}
+
+func (r *secretSyncer) handleSecretEvent(ctx context.Context, event resource.Event[*slim_corev1.Secret]) error {
+	scopedLogger := r.logger.
+		WithField(logfields.K8sNamespace, event.Key.Namespace).
+		WithField("name", event.Key.Name)
+
+	var err error
+
+	switch event.Kind {
+	case resource.Upsert:
+		scopedLogger.Debug("Received Secret upsert event")
+		err = r.handleUpsert(ctx, event.Key, event.Object)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle Secret upsert")
+			err = fmt.Errorf("failed to handle CEC upsert: %w", err)
+		}
+	case resource.Delete:
+		scopedLogger.Debug("Received Secret delete event")
+		err = r.handleDelete(ctx, event.Key)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle Secret delete")
+			err = fmt.Errorf("failed to handle Secret delete: %w", err)
+		}
+	}
+
+	event.Done(err)
+
+	return nil
```

Comment:
Should we return err?  Seems the job controller might report status/metrics based on outcome?

Reply:
good point - of course! thanks!